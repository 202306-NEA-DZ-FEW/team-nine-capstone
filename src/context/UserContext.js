import React, { createContext, useContext, useEffect, useState } from "react";

import { auth } from "@/lib/firebase/controller";

const UserContext = createContext();

export function UserProvider({ children }) {
    const [user, setUser] = useState(null);

    useEffect(() => {
        const isLogged = auth.onAuthStateChanged((authUser) => {
            if (authUser) {
                // User is signed in
                setUser(authUser);
            } else {
                // User is signed out
                setUser(null);
            }
        });

        // Clean up the subscription
        return () => isLogged();
    }, []);

    return (
        <UserContext.Provider value={{ user, setUser }}>
            {children}
        </UserContext.Provider>
    );
}

export function useUser() {
    return useContext(UserContext);
}
