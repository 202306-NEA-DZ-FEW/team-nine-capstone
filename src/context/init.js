import React from "react";

function init() {
    return <div>user context and other stuff</div>;
}

export default init;
