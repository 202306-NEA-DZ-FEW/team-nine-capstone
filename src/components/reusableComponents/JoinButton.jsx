import { arrayRemove, arrayUnion } from "firebase/firestore";
import { useRouter } from "next/router";
import { useTranslation } from "next-i18next";
import React from "react";

import {
    updateEventDocument,
    updateUserDocument,
} from "@/lib/firebase/controller";

import { useUser } from "@/context/UserContext";

function JoinButton({ eventId, eAttendees, setJoinUpdate }) {
    const { t } = useTranslation("common");
    const { user } = useUser();
    const router = useRouter();

    const handleJoin = async () => {
        if (!user) {
            router.push("/authentication/signUp");
        } else {
            // Update event document
            if (eAttendees.includes(user.uid)) {
                // User is in attendees list - remove them
                await updateEventDocument(eventId, {
                    attendees: arrayRemove(user.uid),
                });
                await updateUserDocument(user.uid, {
                    iEvents: arrayRemove(eventId),
                });
            } else {
                // User is not in attendees list - add them
                await updateEventDocument(eventId, {
                    attendees: arrayUnion(user.uid),
                });
                // Event is not in user's events list - add it
                await updateUserDocument(user.uid, {
                    iEvents: arrayUnion(eventId),
                });
            }
        }

        // Update local state
        setJoinUpdate((prev) => prev + 1);
    };

    return (
        <div onClick={handleJoin}>
            {user
                ? eAttendees.includes(user.uid)
                    ? t("joinbtn.joined")
                    : t("joinbtn.join")
                : t("joinbtn.Sign Up")}
        </div>
    );
}

export default JoinButton;
