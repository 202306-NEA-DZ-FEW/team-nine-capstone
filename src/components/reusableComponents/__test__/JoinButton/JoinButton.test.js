import renderer from "react-test-renderer";

import { useUser } from "@/context/UserContext";

import JoinButton from "../../JoinButton";

jest.mock("src/context/UserContext.js");

// Mock the Firebase services
jest.mock("@/lib/firebase/controller.js", () => ({
    getAuth: jest.fn(),
    auth: { currentUser: jest.fn() },
}));

//Mock the Firestore collection and its methods
jest.mock("@/lib/firebase/controller.js", () => ({
    collection: jest.fn(),
}));

// Mock the Firestore instance
jest.mock("@/lib/firebase/controller.js", () => ({
    db: {
        collection: jest.fn(),
    },
}));

describe("JoinButton", () => {
    it("renders correctly", () => {
        // Mock the useUser hook to provide a dummy implementation
        useUser.mockReturnValue({ user: { uid: "dummyUserId" } });

        const tree = renderer.create(<JoinButton />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
