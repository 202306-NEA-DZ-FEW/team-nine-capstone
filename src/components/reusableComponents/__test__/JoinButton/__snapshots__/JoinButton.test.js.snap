// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinButton renders correctly 1`] = `
<div
  onClick={[Function]}
>
  joinbtn.join
</div>
`;
